src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1318,29 +1469,29 @@ public static TSource Single<TSource>(this IEnumerable<TSource> source)
 
         public static TSource Single<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            long count = 0;
-            foreach (TSource element in source)
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (predicate(element))
+                while (e.MoveNext())
                 {
-                    result = element;
-                    checked { count++; }
+                    TSource result = e.Current;
+                    if (predicate(result))
+                    {
+                        while (e.MoveNext())
+                        {
+                            if (predicate(e.Current)) throw Error.MoreThanOneMatch();
+                        }
+                        return result;
+                    }
                 }
             }
-            switch (count)
-            {
-                case 0: throw Error.NoMatch();
-                case 1: return result;
-            }
-            throw Error.MoreThanOneMatch();
+            throw Error.NoMatch();
```

Comment:
With regards to the nested loops, I have the same question/comment about this one as on Last.

Separately, I agree with you that it's unfortunate the current implementation continues iterating the remainder of the list after it's found a second match.  Ideally it wouldn't have been written like that, and ideally we could fix it to not do so.  Changing it, however, is a potentially visible behavioral change.  Such changes need to be discussed to determine whether the compatibility risk is acceptable, and as such they should be done as separate PRs (or potentially an issue discussing it first) that can be evaluated independently.